### ISSUE TITLE
Generating abi after deploying smart contract?

### ISSUE THREAD

**Issue by wangzhen10** [polkadot-js/api#429]
Hi, thanks for providing such a great tool for us. May I ask, do you have a plan to let the SDK can generate abi after deploying smart contract?

**Comment by jacogr** [comment]
Breaking down your question, with the API in-hand it is possible (although long-winded) to interact with a contract. You can generate the call by making use of `api.tx.contracts.call` with the address, gas, value and the parity-codec encoded Bytes data.

The last part of the call is the problematic part and is harder to generate and where the ABI portion will have to come into play. Basically if the types can be passed in, an object that takes care of the encoding can be generated for use.

The short version: yes, we need to provide something, it is a piece of functionality currently missing.

I just want to unpack a bit of what would be useful here. As a contract API user, what do you see - 

1. Passing in an ABI definition (types for calls) and the API uses this to interact
2. A way of generating the ABIs
3. Possibly taking an ABI and generating a JS library
4. Other...

**Comment by wangzhen10** [comment]
Thank you for the reply @jacogr . The 'Passing in an ABI definition (types for calls) and the API uses this to interact' and 'A way of generating the ABIs' are definitely needed. For the 3rd, maybe like web3, generating a contact object by passing the ABI. 

Also, may I ask when are you guys going to develop the ABI feature? And what format are you going to use, the JSON interface like web3?
For example,
`[{
    "type":"function",
    "name":"foo",
    "inputs": [{"name":"a","type":"uint256"}],
    "outputs": [{"name":"b","type":"address"}]
},{
    "type":"event",
    "name":"Event",
    "inputs": [{"name":"a","type":"uint256","indexed":true},{"name":"b","type":"bytes32","indexed":false}],
}]
`

**Comment by jacogr** [comment]
The standard Ethereum ABI JSON format does convey the info, so I'm expecting it to be close to that.  

I don't have a cast-in-stone "when" yet, have been digging through the contracts interfaces to understand the calls (encoding is straight-forward), to have an idea on effort required.

**Comment by jacogr** [comment]
The ABI on the Rust interface is still undergoing some changes as to how methods are identified. (The params encoding follows what is done elsewhere). Initially it was a simple `Compact<U32>` index. 

So for now, until the ABI stabilises and this is more-or-less cast-in-stone, we cannot quite attack this issue yet. (In the Ethereum world the calls are identified by 4 bytes from `keccack(methodName)`, the same does not apply here - elsewhere the identifier is based on the module and function)

(Will link the relevant Substrate issue when it is available)

**Comment by wangzhen10** [comment]
Okay. Thanks for the explanation @jacogr.
